File: 04/Vzorove soubory a ukazky z hodiny/mandelbrot.py
# Python: příprava bodů pro kreslení Mandelbrotovy množiny
# Balíčky Numba pro zrychlení, time pro měření času a sys pro stderr

from numba import njit
xmin=-2; xmax=0.6; ymin=-1.2; ymax=1.2  # obdélník v komplexní rovině
nmax=100                    # max. počet kroků testovací posloupnosti
MaxTime=10                  # doba výpočtu [sec]
output=1                    # 0: jen statistika, 1: data na obrazovku

@njit
def testMandelbrot(x,y):    # výpočet únikového indexu pro bod (x,y)
  c=complex(x,y)
  z=0+0j
  nstop=0
  for n in range(nmax):
    z=z*z+c
    if abs(z)>2:
    # if z.real**2+z.imag**2>4:
      nstop=n+1
      break                 # únikový index nalezen, jinak nmax+1
  return nstop if nstop>0 else nmax+1

import random               # pro náhodná čísla
import time                 # pro měření času
import sys                  # pro výpis na stderr
cnt1=0; cnt2=0              # čítače bodů
tic=time.time()             # startovací čas
while time.time()-tic<MaxTime:        # dokud je doba výpočtu menší než MaxTime
  x=xmin+(xmax-xmin)*random.random()  # náhodné číslo v <xmin,xmax)
  y=ymin+(ymax-ymin)*random.random()  # náhodné číslo v <ymin,ymax)
  nstop=testMandelbrot(x,y)           # únikový index
  if output>0: print('%8.5f%9.5f%5d'%(x,y,nstop))
  # print('{:8.5f} {:8.5f} {:4d}'.format(x,y,nstop))
  # print(f'{x:8.5f} {y:8.5f} {nstop:4d}')
  if nstop==nmax+1: cnt1+=1 # čítač bodů v Mandelbrotově množině 
  cnt2+=1                   # čítač všech prověřených bodů
sys.stderr.write('Mandelbrot: '+str(cnt1)+' / '+str(cnt2)+' = '+f'{100*cnt1/cnt2:5.1f}'+' %\n')
                            # výpis na stderr pomocí write(string), včetně \n pro odřádkování

File: 04/Vzorove soubory a ukazky z hodiny/mandelbrot.f90
! Fortran: příprava bodů pro kreslení Mandelbrotovy množiny
! Modul mTicToc pro měření času v matlabovském stylu tic-toc

include 'mTicToc.f90'           ! vložení obsahu souboru s modulem mTicToc

program Mandelbrot
use mTicToc
implicit none
real :: xmin=-2,xmax=0.6,ymin=-1.2,ymax=1.2 ! obdélník v komplexní rovině
integer :: nmax=100             ! max. počet kroků testovací posloupnosti
real :: MaxTime=10              ! doba výpočtu [sec]
integer:: output=1              ! 0: jen statistika, 1: data na obrazovku
real :: x,y
integer :: nstop,cnt1,cnt2

cnt1=0; cnt2=0                  ! čítače bodů
call tic()                      ! uložení startovacího času
do while (toc()<MaxTime)        ! dokud je doba výpočtu menší než MaxTime
  call random_number(x)         ! náhodné číslo v <0,1)
  call random_number(y)
  x=xmin+(xmax-xmin)*x          ! přepočet do <xmin,xmax)
  y=ymin+(ymax-ymin)*y
  nstop=testMandelbrot(x,y)     ! únikový index
  if (output>0) print '(f8.5,f9.5,i5)',x,y,nstop
  if (nstop==nmax+1) cnt1=cnt1+1 ! čítač bodů v Mandelbrotově množině
  cnt2=cnt2+1                   ! čítač všech prověřených bodů
enddo
write (0,'(a,i0,a,i0,a,f5.1,a)') 'Mandelbrot: ',cnt1,' / ',cnt2,' = ',100.*cnt1/cnt2,' %'
                                ! výpis statistiky na stderr
contains

! Výpočet únikového indexu pro bod (x,y)
integer function testMandelbrot(x,y) result (nstop)
real x,y
complex c,z
integer n
c=cmplx(x,y)
z=(0.,0.)
nstop=0
do n=1,nmax
  z=z*z+c
  if (abs(z)>2) then 
  ! if (z%Re**2+z%Im**2>4) then
    nstop=n
    exit                        ! únikový index nalezen
  endif
enddo
if (nstop==0) nstop=nmax+1      ! únikový index nenalezen, vrátí se nmax+1
end function

end program
